services/inc/util.h

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2017 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <functional>
+#include <type_traits>
+
+namespace particle {
+
+// Wrapper for std::underlying_type that falls back to the original type if it is not an enum
+template<typename T, typename EnableT = void>
+struct UnderlyingType {
+    typedef T Type;
+};
+
+template<typename T>
+struct UnderlyingType<T, typename std::enable_if<std::is_enum<T>::value>::type> {
+    typedef typename std::underlying_type<T>::type Type;
+};
+
+template<typename T>
+inline void combineHash(size_t& seed, const T& value) {
```

Comment:
These look like good additions, but I am always a bit uneasy about 'util" packages/headers etc.  Naming is hard! but let's try to put these  in more tightly-named headers. E.g. `combine_hash.h` for hashing and `underlying_type_enum.h` for the underlying type tweaks.

Reply:
👍